src/JestProcessManagement/JestProcess.ts

```diff
@@ -0,0 +1,66 @@
+import { Runner, ProjectWorkspace } from 'jest-editor-support'
+
+export class JestProcess {
+  private runner: Runner
+  private exited: boolean
+  private projectWorkspace: ProjectWorkspace
+  public keepAlive: boolean
+  public onExitCallback: Function
+  public watchMode: boolean
+
+  private startRunner() {
+    this.exited = false
+    this.runner = new Runner(this.projectWorkspace)
+
+    this.runner.start(this.watchMode)
+
+    this.runner.on('debuggerProcessExit', () => {
+      if (!this.exited) {
+        this.exited = true
+        this.onExitCallback(this)
+        if (this.keepAlive) {
+          this.restart()
+        }
+      }
+    })
+  }
+
+  private restart() {
+    this.runner.removeAllListeners()
+    this.startRunner()
+  }
```

Comment:
should restart() stop the running process before starting?

Reply:
No, I suppose. `restart` is a private method and it is called only internally when process already exited.

Reply:
I see, if it is only used by `startRunner()` and depends on that context, maybe better off making this method an inner function within `startRunner()`, it will prevent accidental misuse and confusion for future maintenance.

Reply:
I do not know... Isn't `private` not communicating that it is...private? What other people think? Could you please give an example of how you would like it to be?

Reply:
it is indeed private, but not for the context of the JestProcess class, but for the `startRunner()` method, because it only makes sense to be invoked after the process is exited as you indicated. Similar to the comment above, we need to be tight about scope/context to avoid confusion and future maintenance overhead.

Suggested change:
```
private startRunner() {
  // move the function here or inline will be fine too
  function restart() {
      this.runner.removeAllListeners()
      this.startRunner()
    }

    this.exited = false
    this.runner = new Runner(this.projectWorkspace)

    this.runner.start(this.watchMode)

    this.runner.on('debuggerProcessExit', () => {
      if (!this.exited) {
        this.exited = true
        this.onExitCallback(this)
        if (this.keepAlive) {
          restart() //here
        }
      }
    })
  }
```

Reply:
one more question, does the keepAlive has a boundary? I have seen sometimes jest can't start, without the upper bound, will the process just keep spawning jest process until it exhausted all system resources?

Reply:
1. In-line functions is not me, but I do not have time to argue about it. I will add your change (and close my eyes when typing :)).
2. Adding a boundary is a good thing. I will add that.

Reply:
> In-line functions is not me

I don't like nested function either, but I dislike out-of-context/running-away-state OOP more ;-)... good news is we don't need a nested function here at all, giving it is only used in 1 place and is only 2 lines, you can further simplify it:
```
private startRunner() {
    let exited = false
    this.runner = new Runner(this.projectWorkspace)

    this.runner.start(this.watchMode)

    this.runner.on('debuggerProcessExit', () => {
      if (!exited) {
        exited = true
        this.onExitCallback(this)
        if (this.keepAlive) {
         // here, even simplier...
          this.runner.removeAllListeners()
          this.startRunner() 
        }
      }
    })
  }
```
hopefully that will keep your eyes open when you type... ;-)